lib/AST/IRToASTVisitor.cpp

```diff
@@ -818,59 +840,75 @@ clang::Stmt *IRToASTVisitor::visitBinaryOperator(llvm::BinaryOperator &inst) {
         ast_ctx.getTypeSize(operand->getType()), sign)};
     return ast.CreateCStyleCast(type, operand);
   }};
+  clang::Expr *res;
   // Where the magic happens
   switch (inst.getOpcode()) {
     case llvm::BinaryOperator::LShr:
-      return ast.CreateShr(IntSignCast(lhs, false), rhs);
+      res = ast.CreateShr(IntSignCast(lhs, false), rhs);
+      break;
 
     case llvm::BinaryOperator::AShr:
-      return ast.CreateShr(IntSignCast(lhs, true), rhs);
+      res = ast.CreateShr(IntSignCast(lhs, true), rhs);
+      break;
 
     case llvm::BinaryOperator::Shl:
-      return ast.CreateShl(lhs, rhs);
+      res = ast.CreateShl(lhs, rhs);
+      break;
 
     case llvm::BinaryOperator::And:
-      return inst.getType()->isIntegerTy(1U) ? ast.CreateLAnd(lhs, rhs)
-                                             : ast.CreateAnd(lhs, rhs);
+      res = inst.getType()->isIntegerTy(1U) ? ast.CreateLAnd(lhs, rhs)
+                                            : ast.CreateAnd(lhs, rhs);
+      break;
 
     case llvm::BinaryOperator::Or:
-      return inst.getType()->isIntegerTy(1U) ? ast.CreateLOr(lhs, rhs)
-                                             : ast.CreateOr(lhs, rhs);
+      res = inst.getType()->isIntegerTy(1U) ? ast.CreateLOr(lhs, rhs)
+                                            : ast.CreateOr(lhs, rhs);
+      break;
 
     case llvm::BinaryOperator::Xor:
-      return ast.CreateXor(lhs, rhs);
+      res = ast.CreateXor(lhs, rhs);
```

Comment:
Perhaps here we should check for XORing against an `llvm::ConstantInt`. If the constant is `ci->isMInusOne()` then emit logical `!` if the size is one, and bitwise `~` is the size isn't one.

Reply:
That's probably something that could be done in a refinement pass